clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4192,6 +4193,142 @@ static bool ContainsWrapperAction(const Action *A) {
   return false;
 }
 
+// Put together an external compiler compilation call which is used instead
+// of the clang invocation for the host compile of an offload compilation.
+void Clang::ConstructHostCompilerJob(Compilation &C, const JobAction &JA,
+                                     const InputInfo &Output,
+                                     const InputInfoList &Inputs,
+                                     const llvm::opt::ArgList &TCArgs,
+                                     const Tool *T) const {
+
+  // The Host compilation step that occurs here is constructed based on the
+  // input from the user.  This consists of the compiler to call and the
+  // options that will be used during the compilation.
+  ArgStringList HostCompileArgs;
+  const InputInfo &InputFile = Inputs.front();
+  const auto &TC = getToolChain();
+
+  // Input file.
+  HostCompileArgs.push_back(InputFile.getFilename());
+
+  // When performing the host compilation, we are expecting to only be
+  // creating intermediate files.  These being preprocessed, assembly or
+  // object files.
+  // We are making assumptions in regards to what options are used to
+  // generate these intermediate files.
+  //                gcc/g++/clang/clang++/default | cl
+  //  Object:                   -c                | -c
+  //  Preprocessed:             -E                | -P -Fi<file>
+  //  Assembly:                 -S                | -c -Fa<file>
+  //  Header Input:        -include <file>        | -FI <file>
+  //
+  // The options used are determined by the compiler name and target triple.
+  auto *HostArg = TCArgs.getLastArg(options::OPT_fsycl_host_compiler_EQ);
+  assert(HostArg && "Expected host compiler designation.");
+
+  bool IsMSVCHostCompiler = false;
+  bool OutputAdded = false;
+  StringRef CompilerName = llvm::sys::path::stem(HostArg->getValue());
+  IsMSVCHostCompiler = CompilerName.contains("cl");
+
+  auto addMSVCOutputFile = [&](StringRef Opt) {
+    SmallString<128> OutOpt(Opt);
+    OutOpt += Output.getFilename();
```

Comment:
Nit:
```suggestion
    SmallString<128> OutOpt(Output.getFilename());
```

Reply:
We would lose the output option with this suggestion.

Reply:
Hm. If I'm not mistaken, my variant would use this constructor:
```cpp
  /// Initialize from a StringRef.
  SmallString(StringRef S) : SmallVector<char, InternalLen>(S.begin(), S.end()) {}
```
which in turn will do
```
  template <typename ItTy,
            typename = std::enable_if_t<std::is_convertible<
                typename std::iterator_traits<ItTy>::iterator_category,
                std::input_iterator_tag>::value>>
  SmallVector(ItTy S, ItTy E) : SmallVectorImpl<T>(N) {
    this->append(S, E);
  }
```

Reply:
What about the input parameter `Opt`?  We need that as well as the `Output.getFilename()` to complete the option name+file

Reply:
Oh, I see. My bad. Please disregard this nit.